### PR TITLE
Validation: Present option to overwrite invalid block

### PR DESCRIPTION
**Pull request by aduth**
Related: #2196, #2132

This pull request seeks to present the user an option to "Overwrite" an invalid block. This is the equivalent of behavior prior to #2132, meaning: The editor will treat the block as it was parsed verbatim, ignoring validation issues.

Overwrite

__Open questions:__

- How can we clarify what "Overwrite" means?

__Testing instructions:__

1. You'll need an invalid block. Try removing a class name from the markup of the first block in `post-content.js`
2. Navigate to Gutenberg > Demo
3. Note the block modified in step 1 is flagged as invalid and you are presented with an "Overwrite" button
4. Press "Overwrite"
5. Note that the warning is dismissed, and the block is revealed as previewed
6. You may also observe that markup has been modified from revisions made in step 1 to the editor's own understanding of the block
   - In my example, I removed the `has-parallax` class from the first Cover Image block, and upon clicking Overwrite, the editor restores the class name because the assigning attribute is parsed from comment attributes

### CODE REVIEW

**Comment by mtias** [comment]
> How can we clarify what "Overwrite" means?

Let's add one more line to the description text with something like "If you don't think there's an issue you can overwrite the changes".

**Comment by youknowriad** [comment]
> How can we clarify what "Overwrite" means?

Could we provide a "preview" of the different recovering options (could be done later)?

File: editor/modes/visual-editor/invalid-block-warning.js [hunk 1]
```diff
@@ -54,6 +62,12 @@ export default connect(
 	null,
 	( dispatch, ownProps ) => {
 		return {
+			ignoreInvalid() {
+				const { block } = ownProps;
+				const serializedBlock = serialize( { ...block, isValid: true } );
+				const parsedBlocks = parse( serializedBlock );
+				dispatch( replaceBlock( block.uid, parsedBlocks ) );
```
Comment (youknowriad) [inline editor/modes/visual-editor/invalid-block-warning.js]: Why can't we just set `isValid` to true? where it's being overridden?

**Review by youknowriad — APPROVED** [review]
Left some minor notes, but this works well

File: editor/modes/visual-editor/invalid-block-warning.js (hunk 1 above)
Comment (aduth) [inline editor/modes/visual-editor/invalid-block-warning.js]: >Why can't we just set `isValid` to true? where it's being overridden?

Yeah, I had considered it. Certainly the most direct approach. It might have been the direct mutation that concerned me. Let me see how it works out.

File: blocks/api/serializer.js [hunk 2]
```diff
@@ -166,11 +166,11 @@ export function serializeBlock( block ) {
...
- * Takes a block list and returns the serialized post content.
+ * Takes a block or set of blocks and returns the serialized post content.
  *
- * @param  {Array}  blocks Block list
+ * @param  {Array}  blocks Block(s) to serialize
  * @return {String}        The post content
  */
 export default function serialize( blocks ) {
-	return blocks.map( serializeBlock ).join( '\n\n' );
+	return castArray( blocks ).map( serializeBlock ).join( '\n\n' );
```
Comment (mtias) [inline blocks/api/serializer.js]: Curious, why is this needed?

File: editor/modes/visual-editor/invalid-block-warning.js (hunk 1 above)
Comment (aduth) [inline editor/modes/visual-editor/invalid-block-warning.js]: Actually, we can't mutate this directly, both because it wouldn't trigger the necessary re-render, because it violates [read-only nature of state](http://redux.js.org/docs/introduction/ThreePrinciples.html#state-is-read-only), and because it wouldn't be detected by selectors which memoize by reference ([example](https://github.com/WordPress/gutenberg/blob/c8a35d4c9be0db263facb8ad9bc1f2220ed1c673/editor/selectors.js)).

File: blocks/api/serializer.js (hunk 2 above)
Comment (aduth) [inline blocks/api/serializer.js]: >Curious, why is this needed?

It's not strictly necessary. In this case the requirement was: being able to serialize a single block. I could have achieved this without any changes to `serialize` by wrapping in an array i.e. `serialize( [ block ] )`, but it occurred to me there's no real reason `serialize` shouldn't be able to be called with a single argument, so this merely normalizes the value.

An alternative approach could also have skipped `map` altogether for singular argument and just `return`ed `serializeBlock`.

Yet another alternative is to expose the individual `serializeBlock` on the block API, and use the specific method appropriate to single / array forms.

**Comment by jasmussen** [comment]
How about this phrasing, does it makes sense?

```
This block looks to have been changed externally. You can keep this change by converting to Classic Text, or you can discard keep it as a [name of block] by fixing it. 

[Discard changes] [Convert and keep changes]
```

And remove the icon. What do you think?

**Comment by aduth** [comment]
>a [name of block]

Using variable substitution in strings can be problematic, an easy example here is difference between "a" and "an", but perhaps also if specific block names could have genders applied depending on the language?

**Comment by aduth** [comment]
>by fixing it

The phrasing is a little odd: Will "discard" fix the block (no), or is the user expected to fix it after discarding?

Discard in general is an odd phrasing. Nothing's being removed or thrown away, it's just that the editor has determined there are conflicts between attributes the block claims to have (either by explicit comment attributes, or reparse) and what would happen if it were to try to save the block with those attributes (differing from what's saved).

**Comment by aduth** [comment]
Ooh, a good opportunity to point to our handy graph:

https://github.com/WordPress/gutenberg/tree/master/editor#editor-an-overview

Basically as soon as the editor loads, we immediately do the entire left side flow once, and expect "Saved Post" _before_ to be (functionally) identical to the "Saved Post" _after_.

**Comment by jasmussen** [comment]
All good thoughts, and I'd appreciate advice on a rephrase that keeps things simple (no need for string substitutions or hoop jumping). The idea is to take some of the tech talk out of it, so it is more easy to parse for someone like me, but without dumbing it down. Tricky I admit.

Restore instead of discard?

**Comment by aduth** [comment]
This is tougher than I thought 😄  In retrospect maybe the "keep" and "discard" wording is good.

- "Keep" in the sense of: I want to keep this block exactly as it is (external modifications and all), and am willing to downgrade it to a base block type, even if it means losing its semantic meaning and editability as a [gallery, cover image, etc]
- "Discard" in the sense of: I want it to remain a [gallery, cover image, etc] even if the external modifications are likely to be overwritten by the editor's attempted understanding of the block.

>This block appears to have been modified externally. Convert to Classic Text to keep your changes, or Discard to overwrite the external changes.
>
>[Convert to Classic Text] [Discard]

**Comment by jasmussen** [comment]
Sounds good to me.

**Comment by mtias** [comment]
Sounds good to me, but I'd put "Discard" first. (We want people to retain the block interactions with which they started, and we want that to be clear.)

File: editor/modes/visual-editor/invalid-block-warning.js (hunk 1 above)
Comment (aduth) [inline editor/modes/visual-editor/invalid-block-warning.js]: I found a slightly better approach here in 9df0a5d, skipping serialize / parse and instead opting to use the `createBlock` API directly (since we have name and attributes of block already).

**Comment by aduth** [comment]
After 858aa59:

Warning

**Comment by mtias** [comment]
Mmm, looking at the image with fresh eyes, it looks like you are discarding the whole block. What about "Continue" instead?

**Comment by aduth** [comment]
"Continue" falls under same problems as "Dismiss" in my opinion, as it makes it sound as a quick exit to the prompt, and not one with any consequence, but in this case there may in fact be consequences to said choice.

**Comment by aduth** [comment]
Going with:

>This block appears to have been modified externally. Overwrite the external changes or Convert to Classic Text to keep your changes.
>
>[ Overwrite ] [ Convert to Classic Text ]
